lib/logstash/inputs/decoder_impl.rb

```diff
@@ -0,0 +1,63 @@
+# encoding: utf-8
+require 'java'
+
+java_import org.logstash.tcp.InputLoop
+java_import org.logstash.tcp.Decoder
+
+class DecoderImpl
+
+  include Decoder
+
+  def initialize(codec, tcp)
+    @tcp = tcp
+    @codec = codec
+    @codecs = {}
+    @first_read = {}
+    @addresses = {}
+    @ports = {}
+    @proxy_addresses = {}
+    @proxy_ports = {}
+  end
+
+  def decode(channel_addr, data)
+    if @codecs[channel_addr].nil?
+      @codecs[channel_addr] = @codec.clone
+      @first_read[channel_addr] = true
+    end
+    bytes = Java::byte[data.readableBytes].new
+    data.getBytes(0, bytes)
+    tbuf = String.from_java_bytes bytes, "ASCII-8BIT"
```

Comment:
@jakelandis the Codec system in Logstash takes care of byte-to-character handling. The Input doesn't need to handle this. 

In Ruby, there is no "byte array" so folks use a String that has encoding ASCII-8BIT which has some nice properties that string indexing/ranges operate on bytes (octets) instead of characters (1-4 (maybe 6?) bytes in UTF-8)